src/audio/OpenALAudioEngine.cpp

```diff
@@ -0,0 +1,485 @@
+         * 
+         * Longer buffers mean lower processing power required,
+         * but higher latency and increased memory usage
+         */
+        static constexpr std::size_t BUFFER_SIZE = 1024;
+    public:
+        StreamingOpenALSound(OpenALAudioEngine* engine,
+                            SoundStream stream, Format fmt, std::size_t sampleRate)
+            : OpenALSound(engine)
+            , m_stream(stream)
+            , m_format(fmt == Format::Mono ? AL_FORMAT_MONO16 : AL_FORMAT_STEREO16)
+            , m_sampleRate(sampleRate)
+            , m_stop(false)
+        {
+            alGenBuffers(BUFFER_NUM, m_buffers.data());
+            ALenum error = alGetError();
+            if (error != AL_NO_ERROR) throw OpenALException(error);
+
+            play();
+
+            m_thread = std::thread(&StreamingOpenALSound::renderThread, this);
+        }
+
+        ~StreamingOpenALSound() override
+        {
+            m_stop = true;
+            m_thread.join();
+            if (m_hasSource)
+            {
+                // Stop source and detach from buffer
+                alSourceStop(m_source);
+                ALint queuedBuffers;
+                alSourceUnqueueBuffers(m_source, BUFFER_NUM, m_buffers.data());
+                alSourcei(m_source, AL_BUFFER, 0);
+                m_engine->m_freeSources.push_back(m_source);
+            }
+            
+            alDeleteBuffers(BUFFER_NUM, m_buffers.data());
+        }
+
+    private:
+        SoundStream m_stream;
+        ALenum m_format;
+        std::size_t m_sampleRate;
+        std::atomic_bool m_stop;
+        std::array<ALuint, BUFFER_NUM> m_buffers{};
+        std::thread m_thread;
+
+        /**
+         * @brief Renders the buffer using the streaming source
+         * 
+         * @param buf The buffer to render data into
+         * @return true The source has more data to render
+         * @return false The source has no more data to render
+         */
+        bool renderBuffer(std::array<std::int16_t, BUFFER_SIZE>& buf)
+        {
+            int rendered = m_stream(buf.data(), BUFFER_SIZE);
+            if(rendered < 0) return false;
+
+            while(rendered < BUFFER_SIZE)
+            {
+                if(m_stop) return false;
+                int res = m_stream(buf.data() + rendered, BUFFER_SIZE - rendered);
+                if(res < 0) return false;
+                rendered += res;
+            }
+            return true;
+        }
+
+        void renderThread()
+        {
+            std::array<std::int16_t, BUFFER_SIZE> dataBuffer{};
+
+            for(auto buf : m_buffers)
+            {
+                if(!renderBuffer(dataBuffer)) return;
+                alBufferData(buf, m_format, dataBuffer.data(), BUFFER_SIZE * 2, m_sampleRate);
+            }
+            
+            alSourceQueueBuffers(m_source, BUFFER_NUM, m_buffers.data());
+
+            if(!renderBuffer(dataBuffer)) return;
+
+            while(true)
```

Comment:
Without further reading it's hard to determine what's the purpose of loop. Some boolean with name explaining what's going on will be useful. ;)